File: TRTC-API-Example-OC/Advanced/CustomCamera/Helper/Render/CustomCameraFrameRender.m
//
//  TestRenderCustomVideoData.m
//  TXLiteAVDemo
//
//  Created by rushanting on 2019/3/27.
//  Copyright © 2019 Tencent. All rights reserved.
//

#import "CustomCameraFrameRender.h"

@interface CustomCameraFrameRender ()
@property (nonatomic, strong) UIImageView* localVideoView;
@property (nonatomic, strong) NSMutableDictionary* userVideoViews;
@property (strong, nonatomic) dispatch_queue_t queue;
@end

@implementation CustomCameraFrameRender

- (instancetype)init
{
    if (self = [super init]) {
        _userVideoViews = [NSMutableDictionary new];
        _queue = dispatch_queue_create(nil, DISPATCH_QUEUE_SERIAL);
    }
    
    return self;
}

- (void)start:(NSString *)userId videoView:(UIImageView *)videoView
{
    // userId is nil for self
    if (!userId) {
        _localVideoView = videoView;
    }
    else {
        [_userVideoViews setObject:videoView forKey:userId];
    }
}
 
- (void)stop {
    UIGraphicsBeginImageContext(_localVideoView.bounds.size);
    UIColor * color = [UIColor clearColor];
    [color setFill];
    UIImage * image = UIGraphicsGetImageFromCurrentImageContext();
    UIGraphicsEndImageContext();
    
    _localVideoView.image = image;
}

- (void)onRenderVideoFrame:(TRTCVideoFrame *)frame userId:(NSString *)userId streamType:(TRTCVideoStreamType)streamType
{
    // When userId is nil, it is the local screen, otherwise it is the remote screen.
    CFRetain(frame.pixelBuffer);
    __weak __typeof(self) weakSelf = self;
    dispatch_async(dispatch_get_main_queue(), ^{
        CustomCameraFrameRender* strongSelf = weakSelf;
        UIImageView* videoView = nil;
        if (userId) {
            videoView = [strongSelf.userVideoViews objectForKey:userId];
        }
        else {
            videoView = strongSelf.localVideoView;
        }
        videoView.image = [UIImage imageWithCIImage:[CIImage imageWithCVImageBuffer:frame.pixelBuffer]];
        videoView.contentMode = UIViewContentModeScaleAspectFit;
        CFRelease(frame.pixelBuffer);
    });
}

@end

File: TRTC-API-Example-Swift/Advanced/CustomCamera/Render/CustomCameraFrameRender.swift
//
//  CustomCameraFrameRender.swift
//  TRTC-API-Example-Swift
//
//  Created by janejntang on 2022/7/4.
//  Copyright © 2022 Tencent. All rights reserved.
//

import Foundation
import UIKit
import TXLiteAVSDK_TRTC
import CoreFoundation
/*
 Custom Video Capturing and Rendering
 TRTC APP supports custom video data collection. This document shows how to customize the data collected by rendering.
 1. Customized rendering of collected data : onRenderVideoFrame()
 For more information, please see https://cloud.tencent.com/document/product/647/34066
 */
class CustomCameraFrameRender:NSObject {
    
    var localVideoView: UIImageView
    let userVideoViews : NSMutableDictionary
    let queue : DispatchQueue
    
    override init() {
        userVideoViews = NSMutableDictionary()
        queue = DispatchQueue(label: "")
        localVideoView = UIImageView()
    }
    
    func start(userId : String?, videoView: UIImageView) {
        if let userID = userId{
            userVideoViews.setObject(videoView, forKey: userID as NSCopying)
        }else{
            localVideoView = videoView
        }
    }
    
    func onRenderVideoFrame(frame: TRTCVideoFrame , userId: String? ,streamType: TRTCVideoStreamType) {
        let pixelBuffer = frame.pixelBuffer
        DispatchQueue.main.async {
            var videoView = UIImageView()
            if let userID = userId{
                guard let videoImageView = self.userVideoViews.object(forKey: userID) as? UIImageView  else { return }
                videoView = videoImageView
            }else{
                videoView = self.localVideoView
            }
            if let pixelBuffer = pixelBuffer {
                videoView.image = UIImage.init(ciImage: CIImage.init(cvImageBuffer: pixelBuffer))
                videoView.contentMode = .scaleAspectFit
            }
        }
    }
    
    func stop() {
        UIGraphicsBeginImageContext(localVideoView.bounds.size)
        let color = UIColor()
        let image = UIGraphicsGetImageFromCurrentImageContext()
        UIGraphicsEndImageContext()
        localVideoView.image = image
    }
}
